RFCs/0051-exclude-operator.adoc

```diff
@@ -0,0 +1,1661 @@
+
+[source,partiql,subs="+{markup-in-source}"]
+----
+-- Original query:
+<select clause>
+EXCLUDE r.s~1~...s~n~
+<from clause>
+<other clauses>
+
+-- Rewritten to:
+<select clause>
+FROM (
+    SELECT VALUE {
+        'r': -- Apply below rewrite rules for steps `s~1~...s~n~`
+        ...  -- Other vars created from the other clauses
+    }
+    <from clause>
+    <other clauses>
+)
+[   -- Include conversion back to list if `ORDER BY` present in `<other clauses>`
+    -- Assume `<topLevelTbl>` and `<idx>` are fresh variables
+    AS <topLevelTbl> AT <idx>
+    ORDER BY <idx>
+]
+----
+
+
+The main idea for rewriting the `EXCLUDE` steps `s~1~,...,s~n~` is to create a nested `CASE` expression for each step, whereby the nested `CASE` expressions for `s~1~,...,s~n-1~` unnest the input binding tuple and the final `CASE` expression for `s~n~` (i.e. the final step) filters out the desired struct field(s) or collection index(es). Every exclude step has an expected type to process during evaluation. Tuple attribute and wildcard exclude steps expect a struct. Whereas a collection index expects a list and a collection wildcard expects a list or bag. The `CASE` expression at each level `i` recreates this expected type by including a `WHEN` branch based on the expected type. Each `CASE` expression will include an `ELSE` branch which outputs the previous level's identifier. This set of branches ensures that at evaluation time, if there is a type mismatch (e.g. evaluation value is a list while the exclude step is a tuple attribute), there is no evaluation error and the previous level's value is returned through the `ELSE` branch. This behavior applies to both the permissive and strict typing modes.
+
+[source,partiql,subs="+{markup-in-source}"]
+----
+-- For the value `r` in our `SELECT VALUE` struct:
+-- Assuming `<v~n-1~>` is the identifier created from the previous exclude step, `s~n-1~`
+SELECT VALUE {
+    'r':
+        CASE
+            WHEN ... --  branch(es) dependent on ``s~1~``'s rewrite rule
+                    ... -- nested `CASE` expressions for `s~2~...s~n~`
+                        CASE
+                            WHEN ... -- branch(es) dependent on ``s~n~``'s rewrite rule
+                        ELSE <v~n-1~>
+                        END
+            ELSE r
+        END
+}
+----
+
+[[anchor-2]]
+.Rewrite rule 2: `EXCLUDE` steps `s~1~,...,s~n-1~`
+For this rewrite rule definition, let `<v~i-1~>` be the identifier created from the previous exclude step (or `r` if this is the first step). For some exclude step `s~i~` that is not the last step, we case on the type of exclude step.
+
+[[anchor-2ai]] Rule 2.a.i::
+    If `s~i~` is a case-sensitive tuple attribute exclude step (e.g. `."foo"` or `['foo']`), where `<v~i~>` and `<attr~i~>` are fresh variables, add the following `WHEN` branch to the `i`^th^ nested `CASE`.
+[source,partiql,subs="+{markup-in-source}"]
+----
+WHEN <v~i-1~> IS STRUCT THEN
+    PIVOT (
```

Comment:
Brought up in a separate conversation w/ Josh, may need to surround the `PIVOT ... FROM UNPIVOT ... ` (or any other subquery) in parens for SQL compatibility. I'll add parens to these subqueries.

Reply:
Resolved in: https://github.com/partiql/partiql-docs/pull/51/commits/3910333bf0c850efafc177e205cecd94c222f764